### PR TITLE
State views

### PR DESCRIPTION
**Pull request by hackaugusto**
These are shortcut functions, used to get state for a given node.

### CODE REVIEW

File: raiden/transfer/views.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+def block_number(node_state):
+    return node_state.block_number
+
+
+def state_from_raiden(raiden):
+    return raiden.wal.state_manager.current_state
+
+
+def state_from_app(app):
+    return app.raiden.wal.state_manager.current_state
```
Comment (palango) [inline raiden/transfer/views.py]: Could be `return state_from_raiden(app.raiden)`

File: raiden/transfer/views.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+
+def get_node_network_status(node_state, node_address):
+    return node_state.nodeaddresses_to_networkstates.get(
+        node_address,
+        NODE_NETWORK_UNKNOWN,
+    )
+
+
+def get_token_network_addresses_for(node_state, payment_network_id):
```
Comment (palango) [inline raiden/transfer/views.py]: What is the `payment_network_id`, the address of the registry contract?

File: raiden/transfer/views.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+def list_channelstate_for_partner(node_state, payment_network_id, partner_address):
+    payment_network = node_state.identifiers_to_paymentnetworks.get(payment_network_id)
+
+    result = []
+    if payment_network is not None:
+
+        for token_network in payment_network.tokenaddresses_to_tokennetworks.values():
+            channel_state = token_network.partneraddresses_to_channels.get(partner_address)
+            if channel_state:
+                # TODO: Either enforce immutability or make a copy
```
Comment (palango) [inline raiden/transfer/views.py]: Does that need to be fixed before merging?

File: raiden/transfer/views.py (hunk 2 above)
Comment (hackaugusto) [reply raiden/transfer/views.py]: yes

File: raiden/transfer/views.py (hunk 3 above)
Comment (hackaugusto) [reply raiden/transfer/views.py]: this wont be a problem as long as the returned object is not mutate, which is not done at the moment

**Comment by palango** [comment]
@hackaugusto a lot of the views could be implemented as properties/methods of the first argument. Is there a reason you decided against that? IMO it makes it easier to discover methods when they're members of the object.

**Comment by hackaugusto** [comment]
The objects that represent the state of the application must only be altered through a state change that is logged, otherwise we can't have recoverability. That means the application state must **not** be poked around, so we should only access data through read-only views (or copies) [please note that at the moment this is conceptual and not enforced].

Additionaly, you must have a reference to an object in order to call a method of it, that means if you want to get all the registered token address from the registry that you need to poke on the structure of the NodeState and PaymentNetworkState, this leaks implementation details and increases coupling among different layers of the sysmte, that is unwished for.

Lastly, I don't agree that free standing functions are any worse than methods, moving the functions from the views module to the `NodeState` class is effectively just a different namespacing, that doesn't increase the "discoverability" of these methods.

**Comment by palango** [comment]
> Additionally, you must have a reference to an object in order to call a method of it, that means if you want to get all the registered token address from the registry that you need to poke on the structure of the NodeState and PaymentNetworkState, this leaks implementation details and increases coupling among different layers of the system, that is unwished for. [This assumes you proposed moving the views to methods into all State objects]

I don't understand the reasoning here. If the first parameter is the explicit `NodeState` or the implicit `self` doesn't make a difference, does it? I understand that coupling might get reduces because stuff is more explicit though.

Another question? Is there a way in python to have immutable views of data?

> Lastly, I don't agree that free standing functions are any worse than methods, moving the functions from the views module to the NodeState class is effectively just a different namespacing, that doesn't increase the "discoverability" of these methods. [This assumes you proposed moving the views functions only to the NodeState object]

I generally find that I get better completion for object members than for function at import time. IOW, the current `views` namespace collects functions from different areas, whereas the object namespaces sort functions depending on "owner" (which I like more)

**Comment by hackaugusto** [comment]
> I don't understand the reasoning here. If the first parameter is the explicit NodeState or the implicit self doesn't make a difference, does it? I understand that coupling might get reduces because stuff is more explicit though.

This is not what I was talking about, please read the text [enclosed by square brackets]. Coupling increases if the methods are distributed across **multiple** State objects.

> I generally find that I get better completion for object members than for function at import time. IOW, the current views namespace collects functions from different areas, whereas the object namespaces sort functions depending on "owner" (which I like more)

True, when the type of the object is know, it works as a filter. I don't think that is a good enough argument for this change, state objects must not be poked around, I don't think the `NodeState` is any special.

Think of State objects as Data Types and not Classes. The code does not use polymorphism anywhere intentionally, so there is not much reason make choices based on the object oriented paradigm.

**Comment by palango** [comment]
> This is not what I was talking about, please read the text [enclosed by square brackets]. Coupling increases if the methods are distributed across multiple State objects.

Got it now. Makes sense.

> True, when the type of the object is know, it works as a filter. I don't think that is a good enough argument for this change, state objects must not be poked around, I don't think the NodeState is any special.

I asked only to understand better why you chose this design. It makes sense as well, there will always be trade-offs.

From my side this is Ok.

**Comment by hackaugusto** [comment]
> Another question? Is there a way in python to have immutable views of data?

Missed this one. Yes, there are multiple approaches:

- Using the same approach as namedtuple, and [inherent from `tuple`](https://github.com/python/cpython/blob/3.7/Lib/collections/__init__.py#L464) [Whooaa, they are not using eval anymore!!!]
- Raising on [`__setattr__`](https://docs.python.org/3/reference/datamodel.html#object.__setattr__) and using [`__new__`](https://docs.python.org/3/reference/datamodel.html#object.__new__) instead of `__init__` (well if we don't inherent from tuple `__init__` works too)
- Using `__setattr__` to implement persintent objects (copy the value, modify it, and return the new updated value)
- Using a library for the above, e.g. [pyrsistent](https://github.com/tobgu/pyrsistent/)

```python
>>> class A:
...   def __init__(self):
...     object.__setattr__(self, 'a', 1)
...   def __setattr__(self, attr, value):
...     raise RuntimeError()
...
>>> A()
<__main__.A object at 0x7fc0927072e8>
>>> A().a
1
>>> A().a = 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 5, in __setattr__
RuntimeError
```

**Pull request merged**